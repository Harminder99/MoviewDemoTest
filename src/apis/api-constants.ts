export const ImageBaseUrl = 'https://image.tmdb.org/t/p/w300/';
export const ApiKeyMovies = 'Your -- Key -- Here';
export const PopularMovies = `/3/movie/popular?api_key=${ApiKeyMovies}`;
