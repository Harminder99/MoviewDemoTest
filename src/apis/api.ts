// api.js
// Import the Axios library for making HTTP requests

import axios from 'axios';

// Create an Axios instance called 'api' with a base URL for your API
const api = axios.create({
  baseURL: 'https://api.themoviedb.org', // Replace with your API base URL
});

// Export the 'api' instance to make it available for use in other parts of your code
export default api;
