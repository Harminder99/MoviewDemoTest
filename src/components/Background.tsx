import * as React from 'react';

import {StyleSheet, View, ViewProps} from 'react-native';

export interface BackgroundProps extends ViewProps {
  // You can define any additional props you want for the container
  backgroundColor?: string;
}

export function Background({children, ...props}: BackgroundProps) {
  return (
    <View style={styles.container} {...props}>
      {children}
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#FFFFFF',
  },
});
