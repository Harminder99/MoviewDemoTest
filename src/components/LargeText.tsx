import * as React from 'react';

import {
  StyleProp,
  StyleSheet,
  Text,
  TextStyle,
  TouchableOpacity,
  View,
  ViewStyle,
} from 'react-native';

export interface LargeTextProps {
  text: String;
  style?: StyleProp<TextStyle> | undefined;
  btnTxtStyle?: StyleProp<TextStyle> | undefined;
  viewStyle?: StyleProp<ViewStyle> | undefined;
  maxCount?: number;
  testID?: string;
}

export function LargeText({
  text,
  style,
  btnTxtStyle,
  viewStyle,
  maxCount = 300,
  testID,
}: LargeTextProps) {
  const [showMore, setShowMore] = React.useState(false);

  const toggleShowMore = () => {
    setShowMore(!showMore);
  };

  const renderText = () => {
    if (showMore) {
      return (
        <Text style={style}>
          {text}
          <TouchableOpacity onPress={toggleShowMore}>
            <Text style={btnTxtStyle}> Show less</Text>
          </TouchableOpacity>
        </Text>
      );
    } else {
      // You can adjust the length of the truncated text as needed.
      const truncatedText = text.slice(0, maxCount);

      return (
        <Text style={[style]}>
          {truncatedText}
          {maxCount < text.length && (
            <TouchableOpacity style={styles.btn} onPress={toggleShowMore}>
              <Text style={[btnTxtStyle]}>... Show more</Text>
            </TouchableOpacity>
          )}
        </Text>
      );
    }
  };

  return (
    <View style={[viewStyle]} testID={testID}>
      {renderText()}
    </View>
  );
}

const styles = StyleSheet.create({
  btn: {
    marginBottom: -3,
  },
});
