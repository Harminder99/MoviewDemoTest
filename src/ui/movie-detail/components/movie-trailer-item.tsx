import * as React from 'react';

import {Image, StyleSheet, Text, View} from 'react-native';

import {background_playlist_color} from '../../../utils/colors';

export function MovieTrailerItem({index}: {index: number}) {
  return (
    <View style={styles.horizontalContainer}>
      <Image
        testID={`play-trailer-${index + 1}`}
        style={styles.playIcon}
        source={require('../../../assets/images/play-circle.png')}
      />
      <Text testID={`trailer-title-${index + 1}`} style={styles.titleTxt}>
        Play Trailer {index + 1}
      </Text>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  horizontalContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    paddingVertical: 10,
    paddingHorizontal: 10,
    marginVertical: 5,
    backgroundColor: background_playlist_color,
  },
  playIcon: {
    width: 20,
    height: 20,
  },
  titleTxt: {
    marginLeft: 10,
  },
});
