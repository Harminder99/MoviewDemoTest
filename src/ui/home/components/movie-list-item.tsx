import * as React from 'react';

import {Dimensions, Pressable, StyleSheet} from 'react-native';

import FastImage from 'react-native-fast-image';
import {ImageBaseUrl} from '../../../apis/api-constants';
import {Movie} from '../../../apis/api-types';

export interface MovieListItemProps {}

const MovieListItem = ({
  item,
  index,
  onPress,
  testID, 
}: {
  item: Movie;
  index: number;
  onPress: (item: Movie) => void;
  testID: string; 
}) => {
  return (
    <Pressable
      style={styles.container}
      onPress={() => onPress(item)}
      testID={testID}>
      <FastImage
        style={styles.item}
        source={{uri: `${ImageBaseUrl}${item.poster_path}`}}
        resizeMode={FastImage.resizeMode.cover}
      />
    </Pressable>
  );
};

export default MovieListItem;

const styles = StyleSheet.create({
  container: {
    width: '50%',
    height: Dimensions.get('screen').width / 1.6,
  },
  item: {
    width: '100%',
    height: '100%',
  },
});
