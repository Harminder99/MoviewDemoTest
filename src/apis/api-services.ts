import {Movies} from './api-types';
import {PopularMovies} from './api-constants';
import api from './api';
import {useQuery} from '@tanstack/react-query';

// this hook will help you get response from Api
export function useMoviesList() {
  return useQuery<Movies, Error>(['movie'], async () => {
    const {data} = await api.get(PopularMovies);
    return data;
  });
}
