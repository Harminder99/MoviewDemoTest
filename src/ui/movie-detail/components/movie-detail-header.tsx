import * as React from 'react';

import {
  Dimensions,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
} from 'react-native';
import {description_color, sub_header} from '../../../utils/colors';

import FastImage from 'react-native-fast-image';
import {ImageBaseUrl} from '../../../apis/api-constants';
import {LargeText} from '../../../components/LargeText';
import {Movie} from '../../../apis/api-types';
import {getResponsiveFontSize} from '../../../utils/typography';

export interface MovieDetailHeaderProps {
  movie: Movie;
}

export function MovieDetailHeader({movie}: MovieDetailHeaderProps) {
  return (
    <View>
      <View style={styles.horizontalContainer}>
        <FastImage
          testID="movie-poster"
          style={styles.imageView}
          source={{uri: `${ImageBaseUrl}${movie?.poster_path}`}}
          resizeMode={FastImage.resizeMode.cover}
        />

        <View style={styles.verticalRatingView}>
          <Text testID="vote-count" style={styles.voteCountTxt}>
            Votes: {movie?.vote_count ?? 0}
          </Text>
          <Text style={styles.voteAvgTxt}>
            Avg.: {movie?.vote_average ?? 0}/10
          </Text>
          <Text style={styles.releaseTxt}>
            Release: {movie?.release_date ?? 0}
          </Text>
          <View style={styles.btnBottomContainer}>
            <TouchableOpacity
              testID="add-to-favorites-button"
              style={styles.btnContainer}>
              <Text style={styles.voteTitleTxt}>Add to Favorites</Text>
            </TouchableOpacity>
          </View>
        </View>
      </View>

      <LargeText
        testID="movie-description"
        style={styles.descriptionTxt}
        btnTxtStyle={styles.showMoreTxt}
        text={movie?.overview ?? ''}
        viewStyle={{marginTop: 20}}
      />

      <Text testID="trailers-section-title" style={styles.trailerTxt}>
        {'TRAILERS'}
      </Text>
      <View style={styles.viewSpace} />
    </View>
  );
}

const styles = StyleSheet.create({
  horizontalContainer: {
    flexDirection: 'row',
  },
  verticalRatingView: {
    flex: 0.8,
    paddingLeft: 20,
  },
  imageContainer: {
    flex: 0.4,
    height: Dimensions.get('screen').width / 2,
  },
  imageView: {
    flex: 0.4,
    height: Dimensions.get('screen').width / 2,
  },
  btnBottomContainer: {
    position: 'absolute',
    bottom: 0,
    left: 20,
    width: '100%',
  },
  btnContainer: {
    width: '100%',
    marginTop: 10,
    alignItems: 'center',
    padding: 15,
    justifyContent: 'center',
    backgroundColor: sub_header,
  },
  voteTitleTxt: {
    fontSize: getResponsiveFontSize(15),
    fontWeight: '500',
    color: '#FFFF',
  },
  voteCountTxt: {
    fontSize: getResponsiveFontSize(18),
    fontWeight: '500',
  },
  voteAvgTxt: {
    fontSize: getResponsiveFontSize(13),
    marginTop: 5,
    fontWeight: '400',
  },
  releaseTxt: {
    fontSize: getResponsiveFontSize(13),
    marginTop: 5,
    fontWeight: '400',
  },
  descriptionTxt: {
    color: description_color,
    fontSize: getResponsiveFontSize(14),
    fontWeight: '500',
  },
  showMoreTxt: {
    fontSize: getResponsiveFontSize(14),
  },
  trailerTxt: {
    marginTop: 20,
    color: description_color,
    fontSize: getResponsiveFontSize(15),
    fontWeight: '500',
  },
  viewSpace: {
    height: 1,
    width: '100%',
    backgroundColor: '#dbdbdb',
    marginVertical: 2,
  },
});
