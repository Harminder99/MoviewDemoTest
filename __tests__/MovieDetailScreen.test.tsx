import {cleanup, render, waitFor} from '@testing-library/react-native';

import { Movie } from '../src/apis/api-types';
import {MovieDetailHeader} from '../src/ui/movie-detail/components/movie-detail-header'; // Import MovieDetailHeader from the same file
import MovieDetailScreen from '../src/ui/movie-detail/movie-detail-screen';
import React from 'react';

// Mock Redux store and useSelector
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
}));

// Mock a sample movie data for testing
const sampleMovie:Movie = {
  title: 'Test Movie Title',
  overview: 'Test Movie Description',
  poster_path: '/test-poster.jpg',
  vote_count: 100,
  adult: true,
  backdrop_path: '',
  genre_ids: [],
  id: 101,
  original_language: '',
  original_title: '',
  popularity: 0,
  release_date: '',
  video: false,
  vote_average: 0,
};

describe('MovieDetailScreen', () => {
  beforeEach(() => {
    // Mock useSelector to return the sample movie data
    jest
      .spyOn(require('react-redux'), 'useSelector')
      .mockImplementation(() => ({
        movie: sampleMovie,
      }));
  });
  afterEach(cleanup);

  it('should render movie title correctly', async () => {
    const {getByTestId} = render(<MovieDetailScreen />);
    const titleElement = getByTestId('movie-title');

    // Wait for the component to finish rendering
    await waitFor(() => {
      const actualText = titleElement.props.children;
      const expectedText = sampleMovie.title;

      expect(actualText).toBe(expectedText);
    });
  });

});

describe('MovieDetailHeader', () => {
  it('renders correctly', () => {
    const tree = render(<MovieDetailHeader movie={sampleMovie} />);
    expect(tree).toMatchSnapshot();
  });
  
  it('should render movie poster correctly', () => {
    const {getByTestId} = render(<MovieDetailHeader movie={sampleMovie} />);
    const posterElement = getByTestId('movie-poster');
    expect(posterElement).toBeTruthy();
  });

  it('should render movie description correctly', async () => {
    const {queryByText, getByTestId} = render(
      <MovieDetailHeader movie={sampleMovie} />,
    );
    await waitFor(() => {
      const expectedText = sampleMovie.overview;
      const descriptionText = queryByText(expectedText);
      expect(descriptionText?.children).toEqual([expectedText]);
    });
  });
});
