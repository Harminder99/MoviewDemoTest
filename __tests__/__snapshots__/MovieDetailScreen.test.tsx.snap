// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetailHeader renders correctly 1`] = `
<View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "flex": 0.4,
            "height": 375,
          },
        ]
      }
    >
      <FastImageView
        defaultSource={null}
        resizeMode="cover"
        source={
          Object {
            "uri": "https://image.tmdb.org/t/p/w300//test-poster.jpg",
          }
        }
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
        testID="movie-poster"
      />
    </View>
    <View
      style={
        Object {
          "flex": 0.8,
          "paddingLeft": 20,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 18,
            "fontWeight": "500",
          }
        }
        testID="vote-count"
      >
        Votes: 
        100
      </Text>
      <Text
        style={
          Object {
            "fontSize": 13,
            "fontWeight": "400",
            "marginTop": 5,
          }
        }
      >
        Avg.: 
        0
        /10
      </Text>
      <Text
        style={
          Object {
            "fontSize": 13,
            "fontWeight": "400",
            "marginTop": 5,
          }
        }
      >
        Release: 
      </Text>
      <View
        style={
          Object {
            "bottom": 0,
            "left": 20,
            "position": "absolute",
            "width": "100%",
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#746A64",
              "justifyContent": "center",
              "marginTop": 10,
              "opacity": 1,
              "padding": 15,
              "width": "100%",
            }
          }
          testID="add-to-favorites-button"
        >
          <Text
            style={
              Object {
                "color": "#FFFF",
                "fontSize": 15,
                "fontWeight": "500",
              }
            }
          >
            Add to Favorites
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "marginTop": 20,
        },
      ]
    }
    testID="movie-description"
  >
    <Text
      style={
        Array [
          Object {
            "color": "#757575",
            "fontSize": 14,
            "fontWeight": "500",
          },
        ]
      }
    >
      Test Movie Description
    </Text>
  </View>
  <Text
    style={
      Object {
        "borderBottomColor": "#56FF",
        "borderBottomWidth": 1,
        "color": "#757575",
        "fontSize": 15,
        "fontWeight": "500",
        "marginTop": 20,
      }
    }
    testID="trailers-section-title"
  >
    TRAILERS
  </Text>
  <View
    style={
      Object {
        "backgroundColor": "#dbdbdb",
        "height": 1,
        "marginVertical": 2,
        "width": "100%",
      }
    }
  />
</View>
`;
